Update: Invoke zoom out view when navigating on inserter pattern tabs.

jorgefilipecosta opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/44585

This PR invokes the zoom-out view when the user is navigating on the inserter patterns tab and closes it if the user goes out of the inserter or moves to another tab.

cc: @jameskoster, @mcsf

jorgefilipecosta committed:
Update: Invoke zoom out view when navigating on inserter pattern tabs.

jameskoster commented:
Thanks for working on this :)

In terms of implementation, I noticed a couple of issues: 

1. If a block is selected, the [+] button doesn't seem to work properly
2. The animation seems very jumpy compared the simply clicking the zoom out button
3. Clicking the Patterns tab in the post editor causes a crash


https://user-images.githubusercontent.com/846565/198984788-85457acf-a1b7-4a2b-9c7f-e858ef17a06d.mp4

Conceptually, it feels a little unexpected to see the zoom behaviour immediately after opening the inserter, before any patterns are actually visible. I wonder if we should try only invoking zoom when you select a pattern category and the secondary panel appears?

jorgefilipecosta commented:
Hi @jameskoster, I updated the PR, and now it only switches to a zoom-out view when previewing a list of patterns, and I fixed the bug on the post editor.
Unfortunately, the patterns preview seems to be rendered as a popover floating, so the zoom-out view ends up below the patterns preview:
![image](https://user-images.githubusercontent.com/11271197/201695746-8d04e853-eaa1-49c4-a40d-d0acefd323f5.png)
This seems suboptimal. I guess to address this, we need to refactor the patterns tabs and remove the popover, which is complex. cc: @youknowriad in case you know a simple way to make this change.

If we go with the refactoring and try to polish this, we may arrive at this result:
![image](https://user-images.githubusercontent.com/11271197/201696981-2f947da6-7bc2-48c7-80a8-9bb6c91699f8.png)
(In the most common scenario where the template/sidebar is open, the user then explores a pattern category.)

It seems like the zoom-out view becomes too narrow, almost like a mobile view. Do you think this would work and it is worth the refactor, or do you think we should not consider the zoom-out view in this case? Or explore other possibilities like closing the template/block sidebar (at the cost of some additional code complexity and maybe unexpected closing of the sidebar by the user).

jameskoster commented:
Thanks for the update. I agree these details make the overall experience a bit awkward.  The patterns panel overlapping the document is definitely sub-optimal given the whole purpose of zooming out is to better facilitate drag/drop 🙈

Refactoring as you suggest would address that issue, but increases the likelihood of the canvas switching between desktop/mobile, which would also be a bit problematic. 

It's made worse when the Inspector is open as you pointed out, and that compounds because clicking to close the Inspector also closes the Inserter.

Auto-closing the Inspector when a patterns category is opened would help because the canvas would retain its size and device preview switching wouldn't occur, something like:

![panels](https://user-images.githubusercontent.com/846565/201902010-1b674d7c-1601-4c7b-b447-1dccaf928bea.gif)

But I'm really not sure if it's worth the technical overhead. 

cc @jasmussen for more thoughts on this.

jasmussen commented:
It would be nice if the flyout could push the canvas inwards, instead of being a popover, and it would be ideal if it avoided invoking any tablet or mobile previews. I think the near-term question is whether this improves what's in trunk or not. I lean towards yes, but I'll need to dive in a bit more. 

How hard would it be to make Jay's suggestion happen? And is there a way to avoid the tablet/mobile invokations?

Thanks for the PR!

richtabor commented:
> Auto-closing the Inspector when a patterns category is opened would help because the canvas would retain its size and device preview switching wouldn't occur

I think there's certainly merit to the idea. Basically an isolated mode for inserting patterns. Wouldn't need list, or edit/select icons. Kinda like Style Book.
